docs/architecture/adr-057-RPC.md

```diff
@@ -0,0 +1,94 @@
+# ADR 057: RPC
+
+## Changelog
+
+- 19-05-2020: created
+
+## Context
+
+Currently the RPC layer of Tendermint is using a variant of the JSON-RPC protocol. This ADR is meant to serve as a pro/con list for possible alternatives and JSON-RPC.
+
+There are currently two options being discussed: gRPC & JSON-RPC.
+
+### JSON-RPC:
+
+JSON-RPC is a JSON-based RPC protocol. Tendermint has implemented its own variant of JSON-RPC which is not compatible with the [JSON-RPC 2.0 specification](https://www.jsonrpc.org/specification).
+
+**Pros:**
+
+- Simple
+- Well-known and well-understood by users and integrators
+- Easy to interface with (HTTP and JSON)
+- Mature, common, and familiar tooling (e.g. `curl` and `jq`)
+- Integrates reasonably well with web infrastructure (proxies, API gateways, service meshes, caches, etc)
+
+**Cons:**
+
+- No schema support
+- RPC clients must be hand-written
+- Streaming not built into protocol
+- Underspecified types (e.g. numbers and timestamps)
+- Tendermint has its own implementation (not standards compliant, maintenance overhead)
+- Stdlib `jsonrpc` package only supports JSON-RPC 1.0, no dominant package for JSON-RPC 2.0
+- Tooling around documentation/specification (e.g. Swagger) could be better
+- JSON data is larger (offset by HTTP compression)
+- Serializing is slower ([~100% marshal, ~400% unmarshal](https://github.com/alecthomas/go_serialization_benchmarks)); insignificant in absolute terms
+
+### gRPC:
+
+gRPC is a high performant RPC framework. It has been battle tested by a large number of users and is heavily relied on and maintained by countless large corporations.
+
+**Pros:**
+
+- Efficient data retrieval for users, lite clients and other protocols
+- Easily implemented in supported languages (Go, Dart, JS, TS, rust, Elixir, Haskell, ...)
+- Can provide a REST API via [gRPC-gateway](https://github.com/grpc-ecosystem/grpc-gateway)
+- Can provide a GraphQL API via [Rejoiner](https://github.com/google/rejoiner)
+- Defined schema with richer type system (Protocol Buffers)
+- Established conventions for forwards- and backwards-compatibility
+- Bi-directional streaming
+- Servers and clients can be autogenerated in many languages (e.g. Tendermint-rs)
+
+**Cons:**
+
+- Users/integrators may not be familiar with gRPC and Protobuf; learning curve
```

Comment:
note: the current RPC has also had and still has user integration issues. When the hub launched, everyone copy-pasted the code from lunie that handled JSON-RPC connections. Lunie made a npm package that handles this now and most everyone is using the copied version or the package. I do see this being the same for gRPC, one or two teams implement it and everyone piggybacks on it. Since Tendermints data structures do not change often (almost never) then all clients would be able to use the same library without needing to fetch and generate proto stubs.

This comment is meant as an input for reviewers when evaluating.